src/codemodder/codemods/base_codemod.py

```diff
@@ -185,38 +185,74 @@ def get_files_to_analyze(
         """
         ...
 
-    def _apply(
+    def _apply_remediation(
         self,
         context: CodemodExecutionContext,
         rules: list[str],
     ) -> None | TokenUsage:
-        if self.provider and (
-            not (provider := context.providers.get_provider(self.provider))
-            or not provider.is_available
-        ):
-            logger.warning(
-                "provider %s is not available, skipping codemod", self.provider
-            )
+        """
+        Applies remediation behavior to a codemod, that is, each changeset will only be associated with a single finging and no files will be written.
+        """
+        if self._should_skip(context):
             return None
+        results: ResultSet | None = self._apply_detector(context)
 
-        if isinstance(self.detector, SemgrepRuleDetector):
-            if (
-                context.semgrep_prefilter_results
-                and self._internal_name
-                not in context.semgrep_prefilter_results.all_rule_ids()
-            ):
-                logger.debug(
-                    "no results from semgrep for %s, skipping analysis",
-                    self.id,
+        if results is not None and not results:
+            logger.debug("No results for %s", self.id)
+            return None
+
+        if not (files_to_analyze := self.get_files_to_analyze(context, results)):
+            logger.debug("No files matched for %s", self.id)
+            return None
+
+        # Do each result independently and outputs the diffs
+        # gather positional arguments for the map
+        resultset_arguments: list[ResultSet | None] = []
+        path_arguments = []
+        if results:
+            for result in results.results_for_rules(rules):
+                # this need to be the same type of ResultSet as results
+                singleton = results.__class__()
```

Comment:
Sorry for such late feedback but it feels like this could be done more cleanly by implementing a class method on the `ResultSet` type:
```python
T = TypeVar("T", bound=Result) # we may already have this somewhere
@classmethod
def from_single_result(cls, result: T) -> Self:
    new = cls()
    cls.add_result(result)
    return new
```
Alternatively there may be a way to make it a method of `Result` instead, I'm not sure which is cleaner. In either case it avoids the introspection and feels like better encapsulation.